src/algebra/lie/weights.lean

```diff
@@ -0,0 +1,253 @@
+@[simp] lemma zero_weight_space_eq_top_of_nilpotent'
+  [lie_algebra.is_nilpotent R L] [is_nilpotent R L M] :
+  weight_space M (0 : L → R) = ⊤ :=
+begin
+  rw [← lie_submodule.coe_to_submodule_eq_iff, lie_submodule.top_coe_submodule],
+  exact infi_max_gen_zero_eigenspace_eq_top_of_nilpotent R L M,
+end
+
+@[simp] lemma coe_weight_space_of_top [lie_algebra.is_nilpotent R L] (χ : L → R) :
+  (weight_space M (χ ∘ (⊤ : lie_subalgebra R L).incl) : submodule R M) = weight_space M χ :=
+begin
+  ext m,
+  simp only [weight_space, lie_submodule.coe_to_submodule_mk, lie_subalgebra.coe_bracket_of_module,
+    function.comp_app, mem_pre_weight_space],
+  split; intros h x,
+  { obtain ⟨k, hk⟩ := h ⟨x, set.mem_univ x⟩, use k, exact hk, },
+  { obtain ⟨k, hk⟩ := h x, use k, exact hk, },
+end
+
+@[simp] lemma zero_weight_space_eq_top_of_nilpotent
+  [lie_algebra.is_nilpotent R L] [is_nilpotent R L M] :
+  weight_space M (0 : (⊤ : lie_subalgebra R L) → R) = ⊤ :=
+begin
+  /- We use `coe_weight_space_of_top` as a trick to circumvent the fact that we don't (yet) know
+    `is_nilpotent R (⊤ : lie_subalgebra R L) M` is equivalent to `is_nilpotent R L M`. -/
+  have h₀ : (0 : L → R) ∘ (⊤ : lie_subalgebra R L).incl = 0, { ext, refl, },
+  rw [← lie_submodule.coe_to_submodule_eq_iff, lie_submodule.top_coe_submodule, ← h₀,
+    coe_weight_space_of_top, ← infi_max_gen_zero_eigenspace_eq_top_of_nilpotent R L M],
+  refl,
+end
+
+/-- Given a Lie module `M` of a Lie algebra `L`, a weight of `M` with respect to a nilpotent
+subalgebra `H ⊆ L` is a Lie character whose corresponding weight space is non-empty. -/
+structure weight :=
+(to_character : lie_character R H)
+(weight_space_ne_bot : weight_space M to_character ≠ ⊥)
+
+instance : has_coe (weight H M) (module.dual R H) := ⟨λ w, (w.to_character : H →ₗ[R] R)⟩
+
+/-- see Note [function coercion] -/
+instance : has_coe_to_fun (weight H M) := ⟨_, weight.to_character⟩
+
+/-- The zero character is a weight for a non-trivial nilpotent Lie module. -/
+instance [nontrivial M] [lie_algebra.is_nilpotent R L] [is_nilpotent R L M] :
+  inhabited $ weight (⊤ : lie_subalgebra R L) M :=
+⟨{ to_character        := 0,
+   weight_space_ne_bot := by
+   { rw [lie_hom.coe_zero, zero_weight_space_eq_top_of_nilpotent], exact top_ne_bot, }, }⟩
+
+end lie_module
+
+namespace lie_algebra
+
+/-- Given a nilpotent Lie subalgebra `H ⊆ L`, the root space of a map `χ : H → R` is the weight
+space of `L` regarded as a module of `H` via the adjoint action. -/
+abbreviation root_space (χ : H → R) : lie_submodule R H L := lie_module.weight_space L χ
+
+@[simp] lemma zero_root_space_eq_top_of_nilpotent [h : is_nilpotent R L] :
+  root_space (⊤ : lie_subalgebra R L) 0 = ⊤ :=
+lie_module.zero_weight_space_eq_top_of_nilpotent L
+
+/-- A root of a Lie algebra `L` with respect to a nilpotent subalgebra `H ⊆ L` is a weight of `L`,
+regarded as a module of `H` via the adjoint action. -/
+abbreviation root := lie_module.weight H L
```

Comment:
Similarly, I would think that for most of the initial theory `is_root` will be more useful. And then at some point, we want a flexible API for root systems and in particular it should play nice with bundled roots of Lie algebras.

Reply:
Exactly right IMHO.